zarrs_filesystem/src/lib.rs

```diff
@@ -270,31 +279,60 @@ impl ReadableStorageTraits for FilesystemStore {
-                            file.read_to_end(&mut buffer)?;
-                            buffer
-                        }
-                        ByteRange::FromStart(_, Some(length)) | ByteRange::Suffix(length) => {
-                            let length = usize::try_from(length).unwrap();
-                            let mut buffer = vec![0; length];
-                            file.read_exact(&mut buffer)?;
-                            buffer
-                        }
+                #[cfg(target_os = "linux")]
+                if enable_direct {
+                    let file_size = file.metadata()?.size();
+                    let ps: usize = page_size::get();
+                    let range = byte_range.to_range(file_size);
+                    let offset = usize::try_from(range.start).unwrap();
+                    let length = usize::try_from(range.end - range.start).unwrap();
+                    if range.end > file_size {
+                        return Err(StorageError::IOError(Arc::new(std::io::Error::new(std::io::ErrorKind::UnexpectedEof, "TODO: To make test pass and match the behavior in the non-direct_io case, requesting length > file size is not permitted"))));
+                    }
+                    let fd = file.as_raw_fd();
+                    let aligned_offset = offset - (offset % ps);
+                    // If the requested length is less than the page size, the aligned_offset could in theory still add on an extra page to the buffer needed to read in the data aligned.
+                    // So offset % ps represents the starting point in the buffer where the requested data is (see the second-to-last line).
+                    // If that point plus length runs over ps, we need another ps worth of data.
+                    let buf_len = if length < ps { ((offset % ps) + length).next_multiple_of(ps) } else { length.next_multiple_of(ps) + ps };
+                    // Because we can't `read_exact` more bytes than are in the file, we need to use these libc calls to do O_DIRECT reading.
+                    let mut buf_ptr: *mut std::ffi::c_void = std::ptr::null_mut();
+                    let ret = unsafe { libc::posix_memalign(&mut buf_ptr, ps,  buf_len) };
+                    assert!(ret == 0, "posix_memalign failed during O_DIRECT reading");
+                    let buf = unsafe { std::slice::from_raw_parts_mut(buf_ptr.cast::<u8>(), buf_len) };
+                    let read_bytes = unsafe { libc::pread(fd, buf_ptr, buf_len, i64::try_from(aligned_offset).unwrap()) };
+                    assert!(read_bytes >= 0, "pread failed during O_DIRECT reading");
+
+                    let start_in_buf = offset - aligned_offset;
+                    let last_bytes = buf[start_in_buf..(start_in_buf + length)].to_vec();
+                    // Free the unused memory
+                    unsafe { libc::free(buf_ptr) };
+                    return Ok(Bytes::from(last_bytes));
```

Comment:
I was thinking something like this, tests seem to pass.

```suggestion

                    // Allocate an aligned output buffer
                    let length = usize::try_from(range.end - range.start).unwrap();
                    let mut aligned_buf = bytes_aligned(length);
                    aligned_buf.resize(aligned_buf.capacity(), 0); // NOTE: Could avoid memset with spare_capacity_mut + unsafe

                    // Seek to the start of the page
                    let page_delta = range.start % (ps as u64);
                    let page_offset = range.start - page_delta;
                    file.seek(SeekFrom::Start(page_offset))?;

                    // Read as many bytes as required, last page can be fewer
                    let length_from_page_offset = usize::try_from(range.end - page_offset).unwrap();
                    file.read_exact(&mut aligned_buf[..length_from_page_offset])?;

                    // Split out the bytes of interest (zero-copy)
                    let buf = aligned_buf.split_off(usize::try_from(page_delta).unwrap()).split_to(length).freeze();
                    return Ok(buf);
```

Reply:
> While that approach is one allocation and avoids a copy, it would leave a lot of unused memory if byte ranges were small relative to the page size

I've got some changes on main so that `get_partial_values_key` can return an iterator now. That would alleviate such a problem as the codec can discard the buffers as it iterates.

I'd suspect `O_DIRECT` reading could benefit quite a bit from byte range coalescing.

Reply:
> Coalescing means what here

I could imagine something like so there are no redundant reads or copies (if the `Bytes` stuff actually works):
1. Find all pages overlapping the byte ranges, coalesce the pages (merge neighbouring)
2. Read the pages
3. Extract the ranges requested

> I am kind of surprised this worked

Me too! Yeah I'm really not sure about the correctness here..

Reply:
No rush at all! This can slot in a minor release for `zarrs_filesystem`

Reply:
It seems my memory did not fail me:

```rust
running 2 tests
test direct_io ... FAILED
test filesystem ... ok

failures:

---- direct_io stdout ----
Error: IOError(Os { code: 22, kind: InvalidInput, message: "Invalid argument" })


failures:
    direct_io

test result: FAILED. 1 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

error: test failed, to rerun pass `-p zarrs_filesystem --test filesystem`
```

Are you seeing this locally with your suggested change as I am?  This is exactly the sort of error I got, cryptically unhelpful and all (although from what I understand, it means you messed something up with O_DIRECT).

Reply:
Ok, this is a real odd one. 04fe231 is passing all tests (repeatedly) on
- My laptop: Arch with ext4 filesystem + kernel 6.16
- Gadi supercomputer: Rocky Linux with lustre filesystem + kernel 4.18

I've not much idea what impacts direct IO behaviour, but clearly we have consistent failure in CI + your cluster. So I think the `posix_memalign` approach will have to do. I think a copy can still be avoided fortunately. Just need a struct owns the allocated buffer, implements `AsRef<[u8]>` and `Drop` -> `libc:free` and then I think [`Bytes::from_owner`](https://docs.rs/bytes/latest/bytes/struct.Bytes.html#method.from_owner) is applicable